src/Juvix/Compiler/Asm/Extra/Type.hs

```diff
@@ -95,7 +103,18 @@ unifyTypes ty1 ty2 = case (ty1, ty2) of
       throw $ AsmError loc ("not unifiable: " <> ppTrace tab ty1 <> ", " <> ppTrace tab ty2)
 
 unifyTypes' :: Member (Error AsmError) r => Maybe Location -> InfoTable -> Type -> Type -> Sem r Type
-unifyTypes' loc tab ty1 ty2 = runReader loc $ runReader tab $ unifyTypes ty1 ty2
+unifyTypes' loc tab ty1 ty2 =
+  runReader loc $
+    runReader tab $
+      if
+          | typeTarget (uncurryType ty1) == TyDynamic && typeTarget (uncurryType ty2) == TyDynamic ->
+              unifyTypes (curryType ty1) (curryType ty2)
```

Comment:
The validation on the JuvixAsm level was too strict for the compilation concept where CallClosures is used to all function types with the dynamic type in the target. Now if both types have dynamic in the (ultimate) target, then they are treated as if they were uncurried (because that's how CallClosures will treat the functions), so e.g. `(A, B) -> A -> *` unifies with `A -> B -> *`.

Reply:
I added a comment